bee-tangle/src/urts.rs

```diff
@@ -0,0 +1,197 @@
+                let mut metadata = TipMetadata::new();
+                metadata.children.insert(child);
+                metadata.time_first_child = Some(Instant::now());
+                entry.insert(metadata);
+            }
+        }
+    }
+
+    fn check_retention_rules_for_parent(&mut self, parent: &MessageId) {
+        // For every tip we add to the pool we call `add_child()`. `add_child()` makes sure that the parents of the tip
+        // are present in the pool. Since `check_retention_rules_for_parent()` will be called after `add_child()` we
+        // can be sure that the parents do exist. Therefore, unwrapping the parents here is fine.
+        if self.non_lazy_tips.len() > MAX_LIMIT_NON_LAZY as usize
+            || self.tips.get(parent).unwrap().children.len() > MAX_NUM_CHILDREN as usize
+            || self
+                .tips
+                .get(parent)
+                .unwrap()
+                .time_first_child
+                .unwrap()
+                .elapsed()
+                .as_secs()
+                > MAX_AGE_SECONDS_AFTER_FIRST_CHILD as u64
+        {
+            self.tips.remove(parent);
+            self.non_lazy_tips.remove(parent);
+        }
+    }
+
+    pub(crate) async fn update_scores<B: StorageBackend>(&mut self, tangle: &MsTangle<B>) {
+        let mut to_remove = Vec::new();
+
+        for tip in self.tips.keys() {
+            match self.tip_score::<B>(tangle, &tip).await {
+                Score::SemiLazy | Score::Lazy => {
+                    to_remove.push(*tip);
+                }
+                _ => continue,
+            }
+        }
+
+        for tip in to_remove {
+            self.tips.remove(&tip);
+            self.non_lazy_tips.remove(&tip);
+        }
+
+        debug!("Non-lazy tips {}", self.non_lazy_tips.len());
+    }
+
+    async fn tip_score<B: StorageBackend>(&self, tangle: &MsTangle<B>, message_id: &MessageId) -> Score {
+        // in case the tip was pruned by the node, consider tip as lazy
+        if !tangle.contains(message_id).await {
+            Score::Lazy
+        } else {
+            let smi = *tangle.get_solid_milestone_index();
+
+            // The tip pool only works with solid tips. Therefore, all tips added to the pool can be considered to
+            // solid. The solid flag will be set together with omrsi and ymrsi values. Therefore, when a
+            // message is solid, omrsi and ymrsi values are available. Therefore, unwrapping here is fine.
+            let omrsi = *tangle.omrsi(&message_id).await.unwrap().index();
+            let ymrsi = *tangle.ymrsi(&message_id).await.unwrap().index();
+
+            if smi > ymrsi + YMRSI_DELTA || smi > omrsi + self.below_max_depth {
+                Score::Lazy
+            } else if smi > omrsi + OMRSI_DELTA {
+                Score::SemiLazy
+            } else {
+                Score::NonLazy
+            }
+        }
+    }
+
+    pub fn two_non_lazy_tips(&self) -> Option<Vec<MessageId>> {
```

Comment:
```suggestion
    pub fn non_lazy_tips(&self) -> Option<Vec<MessageId>> {
```
It doesn't return only two anymore